radicle-cli/src/git.rs

```diff
@@ -175,15 +204,13 @@ pub fn is_signing_configured(repo: &Path) -> Result<bool, anyhow::Error> {
 }
 
 /// Return the list of radicle remotes for the given repository.
-pub fn rad_remotes(repo: &git2::Repository) -> anyhow::Result<Vec<git2::Remote>> {
+pub fn rad_remotes(repo: &git2::Repository) -> anyhow::Result<Vec<Remote>> {
     let remotes: Vec<_> = repo
         .remotes()?
         .iter()
         .filter_map(|name| {
             let remote = repo.find_remote(name?).ok()?;
-            let url = remote.url()?;
-            url.starts_with(radicle::storage::git::transport::local::Url::SCHEME)
-                .then_some(remote)
+            Remote::try_from(remote).ok()
```

Comment:
I think it is done by the `from_str`? I need to take e a look more close but I guess this check is already done

In fact the from_str as this check 

```rust 
 impl FromStr for Url {                                                                                                                                               | 210         .iter()                                                                                                                                                       
  77     type Err = UrlError;                                                                                                                                             | 211         .filter_map(|name| {                                                                                                                                          
  78                                                                                                                                                                      | 212             let remote = repo.find_remote(name?).ok()?;                                                                                                               
  79     fn from_str(s: &str) -> Result<Self, Self::Err> {                                                                                                                | 213             Remote::try_from(remote).ok()                                                                                                                             
  80         let rest = s                                                                                                                                                 | 214         })                                                                                                                                                            
  81             .strip_prefix("rad://")                                                                                                                                  | 215         .collect();                                                                                                                                                   
  82             .ok_or(UrlError::UnsupportedScheme)?;                                                                                                                    | 216     Ok(remotes)                                                                                                                                                       
  83         let components = rest.split('/').collect::<Vec<_>>();                                                                                                        | 217 }                                                                                                                                                                     
  84                                                                                                                                                                      | 218                                                                                                                                                                       
  85         let (resource, namespace) = match components.as_slice() {                                                                                                    | 219 /// Check if the git remote is configured for the `Repository`.                                                                                                       
  86             [resource] => (resource, None),                                                                                                                          | 220 pub fn is_remote(repo: &git2::Repository, alias: &str) -> anyhow::Result<bool> {                                                                                      
  87             [resource, namespace] => (resource, Some(namespace)),                                                                                                    | 221     match repo.find_remote(alias) {                                                                                                                                   
  88             _ => return Err(UrlError::InvalidFormat),                                                                                                                | 222         Ok(_) => Ok(true),                                                                                                                                            
  89         };                                                                                                                                                           | 223         Err(err) if err.code() == git2::ErrorCode::NotFound => Ok(false),                                                                                             
  90                                                                                                                                                                      | 224         Err(err) => Err(err.into()),                                                                                                                                  
  91         let resource = Id::from_canonical(resource).map_err(UrlError::InvalidRepository)?;                                                                           | 225     }                                                                                                                                                                 
  92         let namespace = namespace                                                                                                                                    | 226 }                                                                                                                                                                     
  93             .map(|pk| Namespace::from_str(pk).map_err(UrlError::InvalidNamespace))                                                                                   | 227                                                                                                                                                                       
  94             .transpose()?;                                                                                                                                           | 228 /// Get the repository's "rad" remote.                                                                                                                                
  95                                                                                                                                                                      | 229 pub fn rad_remote(repo: &Repository) -> anyhow::Result<(git2::Remote, Id)> {                                                                                          
  96         Ok(Url {                                                                                                                                                     | 230     match radicle::rad::remote(repo) {                                                                                                                                
  97             repo: resource,                                                                                                                                          | 231         Ok((remote, id)) => Ok((remote, id)),                                                                                                                         
  98             namespace,                                                                                                                                               | 232         Err(radicle::rad::RemoteError::NotFound(_)) => Err(anyhow!(                                                                                                   
  99         })                                                                                                                                                           | 233             "could not find radicle remote in git config; did you forget to run `rad init`?"                                                                          
 100     }                                                                                                                                                                | 234         )),                                                                                                                                                           
 101 }  
```

Reply:
Sorry, maybe I'm not explaining myself clearly. I'm saying that the code should be more like this:
```rust
match Remote::try_from(remote) {
    // ignore URLs that are not rad://
    Err(RemoteError::ParseUrl(UrlError::UnsupportedScheme)) => None,
    // raise an error if there's an invalid URL
    Err(err) => return Err(err), // I think this actually won't work because we're in `filter_map`
    Ok(remote) => Some(remote)
}
```

Note that I'm introducing the `RemoteError` type I suggested creating. 

To get around `filter_map` not being able to handle the `return Err`, I think a better approach is that this function returns:
```rust
impl Iterator<Item = Result<Remote, RemoteError>>
```
that way the caller can decide what to do with the error, e.g. report it and move on, ignore it, re-raise it, etc.

Reply:
Mh, do you mean to return the error to the caller? what is the benefit of doing this if the caller need to filter the code anyway?

What about this

```rust
+            match Remote::try_from(remote) {
+                // ignore URLs that are not rad://
+                Err(RemoteError::ParseUrl(UrlError::UnsupportedScheme)) => Ok(None),
+                // raise an error if there's an invalid URL
+                Err(err) => Err(err), // I think this actually won't work because we're in `filter_map`
+                Ok(remote) => Ok(Some(remote)),
+            }.ok()?
         })
         .collect();

```